modules/host/contractmanager/sectoradd.go

```diff
@@ -0,0 +1,216 @@
+		// this root.
+		location, exists := wal.cm.sectorLocations[id]
+		if exists {
+			// All that needs to happen is the count should be incremented and
+			// updated, and the sectorAdd should be filled out.
+			sa.Count = location.count + 1
+			sa.Folder = location.storageFolder
+			sa.Index = location.index
+
+			// Data can be wiped, as the data is already there, there's no need
+			// to commit to the data or to write it to disk.
+			sa.Data = nil
+
+			// Update the location count to indicate that a sector has been
+			// added.
+			location.count += 1
+			wal.cm.sectorLocations[id] = location
+		} else {
+			// Find a committed storage folder that has enough space to receive
+			// this sector.
+			//
+			// TODO: this search needs to consider storage folders that are in
+			// the process of being removed, and should avoid any storage
+			// folders which are undergoing a remove operation which is
+			// incomplete. This would look like a pre-process which compares
+			// the list of committed storage folders to the list of uncommitted
+			// removals, before passing the result to emptiestStorageFolder.
+			sfs := wal.storageFolders()
+			sf, _ := emptiestStorageFolder(sfs)
+			if sf == nil {
+				// None of the storage folders have enough room to house the
+				// sector.
+				return errInsufficientStorageForSector
+			}
+
+			// Find a location for the sector within the file using the Usage
+			// field.
+			sectorIndex := randFreeSector(sf.Usage)
+
+			// Set the sectorAdd fields and update the sectorLocations map to
+			// reflect this new sector.
+			sa.Count = 1
+			sa.Folder = sf.Index
+			sa.Index = sectorIndex
+			wal.cm.sectorLocations[id] = sectorLocation{
+				index:         sectorIndex,
+				storageFolder: sf.Index,
+				count:         1,
+			}
+		}
+
+		// Add a change to the WAL to commit this sector to the provided index.
+		err := wal.appendChange(stateChange{
+			AddedSectors: []sectorAdd{sa},
+		})
+		if err != nil {
+			return build.ExtendErr("failed to add a state change", err)
+		}
+
+		// Grab the synchronization channel so that we know when the sector add
+		// has completed.
+		syncChan = wal.syncChan
+		return nil
+	}()
+	if err != nil {
+		return build.ExtendErr("verification failed:", err)
+	}
+
+	// Only return after the commitment has succeeded fully.
+	<-syncChan
+	return nil
+}
+
+// commitAddSector will commit a sector that has been added to the WAL. The
+// commit should be idempotent, meaning if this function is run multiple times
+// on the same sector, there should be no issues.
+func (wal *writeAheadLog) commitAddSector(sa sectorAdd) {
+	sf := wal.cm.storageFolders[sa.Folder]
+	lookupTableSize := len(sf.Usage) * 64 * 16
+
+	// Write the sector metadata to disk.
+	writeData := make([]byte, 16)
+	copy(writeData, sa.ID[:])
+	binary.LittleEndian.PutUint16(writeData[12:14], sa.Count)
+	binary.LittleEndian.PutUint16(writeData[14:], sa.Folder)
+	_, err := sf.file.Seek(16*int64(sa.Index), 0)
```

Comment:
`binary.Write` may be appropriate here:
```go
_, err := sf.file.Seek(16*int64(sa.Index), 0)
meta := struct{
    ID [12]byte
    Count uint16
    Folder uint16
}{sa.ID, sa.Count, sa.Folder}
err = binary.Write(sf.file, binary.LittleEndian, meta)
```
Most likely, you would define the `meta` type separately. Then you could use `binary.Read` to decode it.

You could also combine the two `uint16`s into one `uint64` and use the `encoding` package. Then you only waste 4 bytes. I'm not a big fan of bitpacking though